File: #64_Minimum_Path_Sum/Solution.java
public class Solution {
    public int minPathSum(int[][] grid) {
    	if (grid == null || grid.length == 0 || grid[0].length == 0) return 0;
    	int rows = grid.length; int cols = grid[0].length;
    	int[] dp = new int[cols];
    	dp[0] = grid[0][0];
    	/* row 0 */
    	for (int j = 1; j < cols; j++) dp[j] = dp[j - 1] + grid[0][j];
    	/* row 1 to rows - 1 */
    	for (int i = 1; i < rows; i++) {
    		dp[0] += grid[i][0];
    		for (int j = 1; j < cols; j++) {
    			dp[j] = Math.min(dp[j], dp[j - 1]) + grid[i][j];
    		}
    	}
    	return dp[cols - 1];
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[][] t = {
    		{1, 2, 3, 4},
    		{1, 2, 4, 8},
    		{3, 3, 1, 2}
    	};
    	System.out.println(sol.minPathSum(t));
    }
}

File: #64_Minimum_Path_Sum/Solution.py
class Solution:
	# @param {integer[][]} grid
	# @return {integer}
	def minPathSum(self, grid):
		if not grid or not grid[0]:
			return 0
		rows = len(grid)     # rows
		cols = len(grid[0])  # cols
		dp = [[0 for j in range(cols)] for i in range(rows)]
		dp[0][0] = grid[0][0]
		for j in range(1, cols):
			dp[0][j] = grid[0][j] + dp[0][j - 1]
		for i in range(1, rows):
			dp[i][0] = grid[i][0] + dp[i - 1][0]
		for i in range(1, rows):
			for j in range(1, cols):
				dp[i][j] = min(dp[i - 1][j] + grid[i][j], dp[i][j - 1] + grid[i][j])
		return dp[rows - 1][cols - 1]

sol = Solution()
print sol.minPathSum([[1, 2, 5], [3, 2, 1]])